File: scripts/validate_registry.py
from __future__ import annotations

import argparse
import json
from pathlib import Path

from jsonschema import Draft202012Validator


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--registry", type=str, required=True)
    ap.add_argument("--schema", type=str, required=True)
    args = ap.parse_args()

    registry = json.loads(Path(args.registry).read_text(encoding="utf-8"))
    schema = json.loads(Path(args.schema).read_text(encoding="utf-8"))
    validator = Draft202012Validator(schema)

    components = registry.get("components", [])
    failures = []
    for i, comp in enumerate(components):
        errors = sorted(validator.iter_errors(comp), key=lambda e: e.path)
        for e in errors:
            failures.append(f"component[{i}] {comp.get('id')}: {list(e.path)} {e.message}")

    if failures:
        print("Registry validation failed:")
        for f in failures[:50]:
            print(f"- {f}")
        raise SystemExit(1)

    print(f"Registry valid. Components: {len(components)}")


if __name__ == "__main__":
    main()

File: scripts/validate-registry.js
const Ajv = require('ajv');
const fs = require('fs');
const path = require('path');

const ajv = new Ajv({ strict: false });

const schema = JSON.parse(fs.readFileSync(path.join(__dirname, '../schemas/component_spec.schema.json'), 'utf8'));
const registry = JSON.parse(fs.readFileSync(path.join(__dirname, '../data/registry.json'), 'utf8'));

console.log(`Validating ${registry.components.length} components...`);

let validCount = 0;
let errorCount = 0;
const allErrors = [];

registry.components.forEach((component, index) => {
  const validate = ajv.compile(schema);
  const valid = validate(component);
  if (valid) {
    validCount++;
  } else {
    errorCount++;
    allErrors.push({ index: index, id: component.id, errors: validate.errors });
  }
});

console.log('\n=== Validation Results ===');
console.log(`Valid components: ${validCount}`);
console.log(`Invalid components: ${errorCount}`);

if (errorCount > 0) {
  console.log('\n=== Errors ===');
  allErrors.forEach(err => {
    console.log(`\nComponent ${err.index + 1} (${err.id}):`);
    err.errors.forEach(e => console.log('  -', e.message));
  });
}

process.exit(errorCount > 0 ? 1 : 0);
